File: vectorizer.js
// vectorizer.js - Converted from vectorizer.py

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ── CONFIG ───────────────────────────────────────────────────────────────────
const CSV_PATH = path.join(__dirname, 'projects_db.csv');
const TOP_K = 5;
// ──────────────────────────────────────────────────────────────────────────────

function autoDetectColumns(headers) {
    const titleCol = headers.find(h => h.toLowerCase().includes('title')) || null;
    const descCol = headers.find(h => h.toLowerCase().includes('desc') || h.toLowerCase().includes('summary')) || null;
    const categoryCol = headers.find(h => h.toLowerCase().includes('category')) || null;
    
    if (!titleCol || !descCol) {
        throw new Error(`Could not find title/description columns in CSV headers: ${headers}`);
    }
    return [titleCol, descCol, categoryCol];
}

function loadProjects() {
    try {
        const csvData = fs.readFileSync(CSV_PATH, 'utf-8');
        const lines = csvData.split('\n');
        const headers = lines[0].split(',');
        const [titleCol, descCol, catCol] = autoDetectColumns(headers);

        const projects = [];
        for (let i = 1; i < lines.length; i++) {
            if (lines[i].trim()) {
                const values = lines[i].split(',');
                const row = {};
                headers.forEach((header, index) => {
                    row[header] = values[index] || '';
                });
                
                projects.push({
                    title: row[titleCol].trim(),
                    description: row[descCol].trim(),
                    category: catCol ? row[catCol].trim() : 'General'
                });
            }
        }
        return projects;
    } catch (error) {
        console.error('Error loading projects:', error);
        return [];
    }
}

function tokenize(text) {
    // lowercase alphanumeric words
    return text.toLowerCase().match(/\w+/g) || [];
}

// ── Cold‐start: load & preprocess once ─────────────────────────────────────────
const PROJECTS = loadProjects();
const PROJECT_TOKENS = PROJECTS.map(p => [
    tokenize(p.title + ' ' + p.description), 
    p
]);
// ──────────────────────────────────────────────────────────────────────────────

export function getSimilarProjects(query, topK = TOP_K) {
    /**
     * Score each project by how many words from the query appear in its
     * title+description, then return the top_k best matches.
     */
    const qTokens = tokenize(query);
    const scored = [];
    
    for (const [tokens, project] of PROJECT_TOKENS) {
        const score = qTokens.filter(qt => tokens.includes(qt)).length;
        scored.push([score, project]);
    }

    // sort high→low
    scored.sort((a, b) => b[0] - a[0]);

    // take those with nonzero overlap first...
    let results = scored.filter(([score, _]) => score > 0).map(([_, proj]) => proj).slice(0, topK);
    
    // ...then pad with next-best if we don't have enough
    if (results.length < topK) {
        const remaining = scored.filter(([_, proj]) => !results.includes(proj)).map(([_, proj]) => proj);
        results = results.concat(remaining.slice(0, topK - results.length));
    }
    
    return results;
} 

File: vectorizer.py
# api/vectorizer.py

import csv, os, re

# ── CONFIG ───────────────────────────────────────────────────────────────────
CSV_PATH = os.path.join(os.path.dirname(__file__), 'projects_db.csv')
TOP_K    = 5
# ──────────────────────────────────────────────────────────────────────────────

def auto_detect_columns(headers):
    title_col    = next((h for h in headers if 'title' in h.lower()), None)
    desc_col     = next((h for h in headers if 'desc' in h.lower() or 'summary' in h.lower()), None)
    category_col = next((h for h in headers if 'category' in h.lower()), None)
    if not title_col or not desc_col:
        raise Exception(f"Could not find title/description columns in CSV headers: {headers}")
    return title_col, desc_col, category_col

def load_projects():
    with open(CSV_PATH, newline='', encoding='utf-8') as f:
        reader  = csv.DictReader(f)
        headers = reader.fieldnames or []
        title_col, desc_col, cat_col = auto_detect_columns(headers)

        projects = []
        for row in reader:
            projects.append({
                'title':       row.get(title_col, '').strip(),
                'description': row.get(desc_col,   '').strip(),
                'category':    row.get(cat_col,    'General').strip() if cat_col else 'General'
            })
    return projects

def tokenize(text):
    # lowercase alphanumeric words
    return re.findall(r'\w+', text.lower())

# ── Cold‐start: load & preprocess once ─────────────────────────────────────────
_projects       = load_projects()
_project_tokens = [
    ( tokenize(p['title'] + ' ' + p['description']), p )
    for p in _projects
]
# ──────────────────────────────────────────────────────────────────────────────

def get_similar_projects(query: str, top_k: int = TOP_K):
    """
    Score each project by how many words from the query appear in its
    title+description, then return the top_k best matches.
    """
    q_tokens = tokenize(query)
    scored   = []
    for tokens, project in _project_tokens:
        score = sum(1 for qt in q_tokens if qt in tokens)
        scored.append((score, project))

    # sort high→low
    scored.sort(key=lambda x: x[0], reverse=True)

    # take those with nonzero overlap first...
    results = [proj for score, proj in scored if score > 0][:top_k]
    # ...then pad with next-best if we don’t have enough
    if len(results) < top_k:
        results += [proj for _, proj in scored if proj not in results][: top_k - len(results)]
    return results
